### PR TITLE
[Feature] Color-blind themes

### PR DESCRIPTION
**Pull request by prune998**
# Description

This Pull Request adds 3 more default themes for dark terminal (+3 light versions), based on the 3 main color-blindness disabilites : deuteranopia, tritanopia and protanopia.

While they are pretty basic themes, and all themes shares the same presets, they will be enhanced over time when we work closer with the community and gets feedbacks.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## What you changed
Addition of 6 defaut themes, documentation.

- also added a `themes` folder with an example theme to help users bootstrap their own.
- also added some `kubernetes` yaml files to greate resources in a cluster to test different aspects of kubecolor (still needs documentation)

### CODE REVIEW

File: kubecolor.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+class Kubecolor < Formula
```
Comment (applejag) [inline kubecolor.rb:1]: This file feels out of place.

Accidentally committed? Or why is it here?

File: themes/my-test-theme-dark/README.md [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+If you want to share your custom theme, please open a Pull Request and:
+- explain in the PR what is the goal of the theme (Ex: simple esthetic, disability related, better efficiency to display informations,...)
+- add a foder for your theme into the `themes` folder, and name it with name of your theme and either the `-dark` or `-light` suffix. Ex: `my-test-theme-dark`.
+- a screen capture of the end result as `.jpg` or `.png`
+- the `color.yaml` file that includes comments with:
+  - your name
+  - your email
+  - the date of the last update
+
+As maintainers we reserve the right to refuse any theme, whatever our reason is. If you don't agree, you are free to host your theme anywhere else.
```
Comment (applejag) [inline themes/my-test-theme-dark/README.md:12]: Feels a bit hostile in the phrasing.

I think this is a given and could be skipped


```suggestion
```

File: themes/my-test-theme-dark/README.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# KubeColor Themes
+
+If you want to share your custom theme, please open a Pull Request and:
+- explain in the PR what is the goal of the theme (Ex: simple esthetic, disability related, better efficiency to display informations,...)
+- add a foder for your theme into the `themes` folder, and name it with name of your theme and either the `-dark` or `-light` suffix. Ex: `my-test-theme-dark`.
```
Comment (applejag) [inline themes/my-test-theme-dark/README.md:5]: Typo


```suggestion
- add a folder for your theme into the `themes` folder, and name it with name of your theme and either the `-dark` or `-light` suffix. Ex: `my-test-theme-dark`.
```

File: themes/my-test-theme-dark/README.md [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+- a screen capture of the end result as `.jpg` or `.png`
```
Comment (applejag) [inline themes/my-test-theme-dark/README.md:6]: I think having a sample directory tree as a comment could be useful.

Such as the output of the `tree` command in a code block

File: themes/my-test-theme-dark/README.md [hunk 5]
```diff
@@ -0,0 +1,12 @@
+# KubeColor Themes
```
Comment (applejag) [inline themes/my-test-theme-dark/README.md:1]: Suggest moving this file to one dir up: `themes/README.md`

And then have the `themes/my-test-theme-dark/README.md` only contain the info you request

File: themes/my-test-theme-dark/color.yaml [hunk 6]
```diff
@@ -0,0 +1,58 @@
+# This is a sample theme used for demo
+# author: Prune <prune@lecentre.net>
+# last update: 20240324
```
Comment (applejag) [inline themes/my-test-theme-dark/color.yaml:3]: Suggest using dashes


```suggestion
# last update: 2024-03-24
```

File: themes/my-test-theme-dark/color.yaml [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+kubectl: kubectl # path to kubectl executable
+preset: dark # color theme preset
+objFreshThreshold: 1m # ages below this uses theme.data.durationfresh coloring
+
```
Comment (applejag) [inline themes/my-test-theme-dark/color.yaml:8]: Suggest removing configs that use the defaults. To keep the theme minimal.

Also, suggest removing the comments as we don't want to maintain these comments too when we update the docs. Instead only use YAML comments to explain stuff that's specific to this theme


```suggestion
objFreshThreshold: 1m

```

Same goes for the rest of this file. Don't want to spam suggestion comments

File: config/theme.go [hunk 8]
```diff
@@ -108,6 +108,161 @@ func NewBaseTheme(preset Preset) *Theme {
...
+			Default: MustParseColor("white"),
+			Base: ThemeBase{
+				Key:       MustParseColorSlice("#feb927 / white"),
+				Info:      MustParseColor("white"),
+				Primary:   MustParseColor("#4860e6"),             // magenta
+				Secondary: MustParseColor("#2aabee"),             // cyan
+				Success:   MustParseColor("#6afd6a:bold"),        // bold green
+				Warning:   MustParseColor("#feb927:italic"),      // yellow
+				Danger:    MustParseColor("fg=white:bg=#c2270a"), // red background
+				Muted:     MustParseColor("#feb927"),             // yellow
```
Comment (applejag) [inline config/theme.go:123]: Maybe we want to set a different color for the muted instead for using the same as the warning color.

I set that for the default theme just to keep backwards compatibility for now, but I think we can have a separate color here for new themes.

File: config/theme.go [hunk 9]
```diff
@@ -108,6 +108,161 @@ func NewBaseTheme(preset Preset) *Theme {
...
+			Data: ThemeData{
+				String: MustParseColor("#2aabee"),
+			},
+			Table: ThemeTable{
+				Header:  MustParseColor("white:bold"),
+				Columns: MustParseColorSlice("#2aabee / #6afd6a:bold / #4860e6 / white / #feb927"),
+			},
+			Options: ThemeOptions{
+				Flag: MustParseColor("#feb927"),
+			},
```
Comment (applejag) [inline config/theme.go:134]: Suggest just using the default fallback color here of secondary.

File: config/theme.go [hunk 10]
```diff
@@ -108,6 +108,161 @@ func NewBaseTheme(preset Preset) *Theme {
 			},
 		}
 
+	// Special Preset for Protanopias
+	case PresetProtDark:
+		return &Theme{
+			Default: MustParseColor("white"),
+			Base: ThemeBase{
+				Key:       MustParseColorSlice("#feb927 / white"),
```
Comment (applejag) [inline config/theme.go:116]: I think it's a bad idea to use white coloring in the keys. It makes it confused with the text color of the values in output like describe.

What if you had a weaker version of `#feb927` as the second key color?

**Review by applejag — CHANGES_REQUESTED** [review]

File: kubecolor.rb (hunk 1 above)
Comment (prune998) [reply kubecolor.rb:1]: removed.

File: themes/my-test-theme-dark/README.md (hunk 4 above)
Comment (prune998) [reply themes/my-test-theme-dark/README.md:6]: I think I added that now... please check

File: themes/my-test-theme-dark/README.md (hunk 5 above)
Comment (prune998) [reply themes/my-test-theme-dark/README.md:1]: I think we don't need a readme there, only comments in the theme file ?

File: themes/my-test-theme-dark/color.yaml (hunk 7 above)
Comment (prune998) [reply themes/my-test-theme-dark/color.yaml:8]: OK, done.

I'll look into the `go run ./internal/cmd/configdoc` thing... maybe we should have an "empty theme" file with all the current values that someone could use a boilerplate, even if they don't have `go` installed ?

File: config/theme.go (hunk 8 above)
Comment (prune998) [reply config/theme.go:123]: OK I set it to something yellowish light, that is green-ish when you're not color-blind... I guess we will adjust when we have some reviews...

Remember that what we see may be slightly strange, but this theme is not meant for normal use.

File: config/theme.go (hunk 9 above)
Comment (prune998) [reply config/theme.go:134]: Set to the secondary "cyan" color

File: config/theme.go (hunk 10 above)
Comment (prune998) [reply config/theme.go:116]: OK, changed
image

File: themes/my-test-theme-dark/README.md (hunk 5 above)
Comment (applejag) [reply themes/my-test-theme-dark/README.md]: Yep you're right. Even simpler

File: themes/my-test-theme-dark/color.yaml (hunk 7 above)
Comment (applejag) [reply themes/my-test-theme-dark/color.yaml]: I think relying on the config docs in the README.md (in the repo root) should hopefully suffice.

File: config/theme.go (hunk 10 above)
Comment (applejag) [reply config/theme.go]: That looks so much better IMO

File: README.md [hunk 11]
```diff
@@ -445,12 +445,21 @@ export KUBECOLOR_THEME_TABLE_COLUMNS="bg=red:underline/bg=green:italic/bg=blue:b
...
+| `deuteranopia-dark`(*)  | Color theme for [Deuteranopia color blindness](https://www.color-blindness.com/deuteranopia-red-green-color-blindness/). For when your terminal's background is dark.
+| `deuteranopia-light`(*) | Color theme for [Deuteranopia color blindness](https://www.color-blindness.com/deuteranopia-red-green-color-blindness/). For when your terminal's background is light/bright.
+| `tritanopia-dark`(*)    | Color theme for [Tritanopia color blindness](https://www.color-blindness.com/tritanopia-blue-yellow-color-blindness/). For when your terminal's background is dark.
+| `tritanopia-light`(*)   | Color theme for [Tritanopia color blindness](https://www.color-blindness.com/tritanopia-blue-yellow-color-blindness/). For when your terminal's background is light/bright.
+
+> [!NOTE]
+> As of version `v0.3.0`, both `deuteranopia` and `tritanopia` themes are the same as `protanopia`. They may differ in future versions when we better tune them. Set your configuration to match your color-blindness type so you will benefit of the futur changes.
```
Comment (applejag) [inline README.md:462]: nitpick: Suggest adding the `(*)` here as well to make it clear that this note is a footnote to the `(*)` part

```suggestion
> (*) As of version `v0.3.0`, both `deuteranopia` and `tritanopia` themes are the same as `protanopia`. They may differ in future versions when we better tune them. Set your configuration to match your color-blindness type so you will benefit of the future changes.
```

File: themes/my-test-theme-dark/README.md (hunk 2 above)
Comment (applejag) [reply themes/my-test-theme-dark/README.md]: I see you added a confused smilie, so I just want to explain myself a little. I specifically meant the last part:

> If you don't agree, you are free to host your theme anywhere else.

Like, I agree with the sentiment, but phrasing that is so tricky without sounding nonchalant (and as I don't have any better suggestion in phrasing it) so I find _not saying it_ is better.

File: themes/README.md [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+```
+themes
+├── README.md
+├── my-test-theme-dark
+│   ├── color.yaml
+│   └── image.png
+└── new-theme-dark
+    ├── color.yaml
+    └── image.png
+```
```
Comment (applejag) [inline themes/README.md:25]: Maybe add a hint in here about how to capture the screenshot.

Such as:

````suggestion
```

## Creating screenshot

Requires that you have kubectl and kubecolor installed,
as well as access to a Kubernetes cluster (e.g via Docker Desktop or Minikube).

Apply the files found in [../test-cluster](../test-cluster), like so:

```bash
kubectl create -f ./test-cluster
```

Then take a screenshot of your terminal.
When you're done, you can delete the resources again:

```bash
kubectl delete -f ./test-cluster
```
````

**Review by applejag — CHANGES_REQUESTED** [review]
Just some additional small comments.

This PR looks great btw! Nice job!

File: themes/my-test-theme-dark/color.yaml [hunk 13]
```diff
@@ -0,0 +1,25 @@
+# This is a sample theme used for demo
```
Comment (applejag) [inline themes/my-test-theme-dark/color.yaml:1]: Oh also could we add the YAML language server comment here too?

```suggestion
# yaml-language-server: $schema=https://github.com/kubecolor/kubecolor/raw/main/config-schema.json

# This is a sample theme used for demo
```

File: themes/my-test-theme-dark/image.png
Comment (applejag) [inline themes/my-test-theme-dark/image.png:1]: This image does not match the sample color theme.

The image has header with blue background, but that's missing from the color.yaml. And the 1/1 ratio should be green according to the color.yaml but in the screenshot its cyan

Comment (prune998) [reply themes/my-test-theme-dark/image.png:1]: I made that too quickly :)
correcting it

File: themes/my-test-theme-dark/color.yaml (hunk 13 above)
Comment (prune998) [reply themes/my-test-theme-dark/color.yaml:1]: It's done, but there's an issue with the `data.theme.durationFresh` that is `durationFresh` or `durationfresh`... It's just json validation for now so we'll sort out later. I need to merge this as I'll be unavailable soon.

Comment (applejag) [reply themes/my-test-theme-dark/color.yaml:3]: Well when viper reads yaml it does so in a case insensitive way. So it shouldn't matter much

**Review by applejag — APPROVED** [review]
LGTM!

We can fix the failing GH Action in a follow up PR. If you were busy then I can look into it

**Pull request merged**